woocommerce/woocommerce
# Fix stock status saving for variable products on bulk and quick edit

### All Submissions:


### Changes proposed in this Pull Request:

Fixes related to stock status for variable products on bulk and quick edit.

* **For bulk edit:** Even if stock status was left as "No change", the status of all variations was being changed to whatever the status of the product was before it was converted to variable. Now no change is performed when "No change" is selected, and all variations change to whatever is selected otherwise.

* **For quick edit:** A new "No change" option is added to the "Stock status" selector that will be preselected when the product is variable. Previously, whatever status the product had before being converted to variable was being shown, and that's the status that would be set when saving. Also, a "This will change the stock status of all variations" message is displayed before the selector.

![image](https://user-images.githubusercontent.com/937723/79217826-c58ba180-7e4f-11ea-8196-b7e6b3ec1a87.png)

#### Other improvements introduced:

* `WC_Unit_Test_Case` class gets two new utility methods: `login_as_role` and `login_as_administrator`.

* `WC_Helper_Product::create_simple_product` gets a new optional `$props` argument.

Closes #25843.

### How to test the changes in this Pull Request:

1. Create a variable product with two variations. Set stock status to "Out of stock" for one of them and to "On backorder" for the other.
2. Bulk edit the product, don't change anything and save. Verify that the stock status of the variations hasn't changed (without the fix, both would have changed to "In stock").
3. Bulk edit the product again, se the stock status selector to "In stock" and save. Verify that this time the status of both variations has changed.
4. Change the stock status of both variations to anything but "In stock" again.
5. Quick edit the product. Verify that the preselected value for the stock status selector is now "No change". Leave it that way and save. Verify that the stock status of the variations hasn't changed.
6. Quick edit the product again. Select "In stock" in the stock status selector. Verify that this time the status of both variations has changed.

### Other information:


### Changelog entry

> Fix - don' change the stock status of variations when bulk editing a variable product and leaving the "Stock status" selector as "No change".
> Enhancement - Add a "No change" option to the "Stock status" selector in quick edit, preselect it when the product being edited is a variable product.
